src/mesh/python/mesh_types.py

```diff
@@ -561,42 +556,72 @@ def norm(v):
-        self.num_faces = len(ridge_vertices)
         self.num_faces_per_cell = np.zeros(self.num_cells, dtype=int)
         for n in range(self.num_cells):
             self.num_faces_per_cell[n] = len(cells[n])
+        self.num_faces = sum(self.num_faces_per_cell)
         self.num_nodes_per_face = np.zeros(self.num_faces, dtype=int)
         for n in range(self.num_faces):
             self.num_nodes_per_face[n] = 2
         self.faces_per_cell = cells
         self.nodes_per_face = ridge_vertices
+
+        # -- rewrite cells and faces to not have duplicate faces
+        new_cells = []
+        new_faces = []
+        for cell in cells:
+            new_cell = []
+            for face in cell:
+                new_faces.append(ridge_vertices[face])
+                new_cell.append(len(new_faces) - 1)
+            new_cells.append(new_cell)
+        self.faces_per_cell = new_cells
+        self.nodes_per_face = new_faces
+
+        # -- write boundaries for new faces
+        boundary_edges = {}
+        boundary_edges['xl'] = []
+        boundary_edges['xr'] = []
+        boundary_edges['yl'] = []
+        boundary_edges['yr'] = []
+        for face_idx, v_indices in enumerate(self.nodes_per_face):
+            midpoint = [(vertices[v_indices[0]][0] + vertices[v_indices[1]][0]) / 2,
+                        (vertices[v_indices[0]][1] + vertices[v_indices[1]][1]) / 2]
+            distances = np.zeros(len(cell_nodes))
+            for n, node in enumerate(cell_nodes):
+                distances[n] = np.sqrt((points[n][0] - midpoint[0])**2 +
+                                       (points[n][1] - midpoint[1])**2)
+            indices = np.argsort(distances)
+            distances = np.sort(distances)
+            if not soft_equiv(distances[0], distances[1]):
+                # -- a boundary
+                if (soft_equiv(vertices[v_indices[0]][0], xmin) and
+                    soft_equiv(vertices[v_indices[1]][0], xmin)):
+                    boundary_edges['xl'].append(face_idx)
+                elif (soft_equiv(vertices[v_indices[0]][0], xmax) and
+                      soft_equiv(vertices[v_indices[1]][0], xmax)):
+                    boundary_edges['xr'].append(face_idx)
+                elif (soft_equiv(vertices[v_indices[0]][1], ymin) and
+                      soft_equiv(vertices[v_indices[1]][1], ymin)):
+                    boundary_edges['yl'].append(face_idx)
+                elif (soft_equiv(vertices[v_indices[0]][1], ymax) and
+                      soft_equiv(vertices[v_indices[1]][1], ymax)):
+                    boundary_edges['yr'].append(face_idx)
+                else:
+                    assert (False), 'Boundary edge not identified'
         self.nodes_per_side = []
         for n in range(4):
             bdy_key = list(boundary_edges.keys())[n]
             bdy_nodes = []
             for bdy in boundary_edges[bdy_key]:
-                nodes = ridge_vertices[bdy]
```

Comment:
Why not use `np.sort`/`np.unique` to remove duplicate nodes?

Reply:
fixed (I think I missed where the `np.sort` should be used though)

Reply:
I think this question was born of my ignorance on the vertex/node terminology - is an `np.unique` actually needed here, or should the node list be unique by construction?

Reply:
I'm looping over faces here when I collect all the boundary nodes/vertices and nodes/vertices here will be duplicated if two faces connect i.e. share a node/vertex. Basically I think it's just that the logic in my script is backwards from the file -- I determine boundaries via faces, but the file provides boundaries as sets of nodes/vertices.

Reply:
Ah ok - then it is indeed useful, at least in compactifying the boundary files.